File: app/policy_engine.py
"""
Policy Enforcement Engine for CmdBar Python backend and companion utilities.
Enforces security policies, MFA for sensitive operations, Data Loss Prevention (DLP),
Geographic restrictions (Geo-fencing), and Time-based access controls.
"""

import re
import time
import base64
import hmac
import hashlib
import struct
import datetime
import ipaddress

DEFAULT_POLICY_CONFIG = {
    "enabled": True,
    "mfa": {
        "enabled": True,
        "sensitive_keywords": [
            "deploy",
            "delete",
            "destroy",
            "drop",
            "sudo",
            "prod",
            "admin",
            "kms",
            "iam",
            "secrets",
        ],
        "session_ttl_seconds": 300,
        "secret": "JBSWY3DPEHPK3PXP",  # Default base32 test key (Secret: "Hello!")
    },
    "dlp": {
        "enabled": True,
        "action": "block",  # "block" | "redact" | "warn"
        "patterns": [
            {
                "name": "AWS Access Key",
                "regex": r"AKIA[0-9A-Z]{16}",
                "category": "credentials",
            },
            {
                "name": "Private Key",
                "regex": r"-----\s*BEGIN[ A-Z1-9_-]*PRIVATE KEY\s*-----",
                "category": "crypto",
            },
            {
                "name": "SSN",
                "regex": r"\b\d{3}-\d{2}-\d{4}\b",
                "category": "pii",
            },
            {
                "name": "Credit Card",
                "regex": r"\b(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13})\b",
                "category": "financial",
            },
            {
                "name": "Generic Secret",
                "regex": r"(?i)(api[_-]?key|secret|password|bearer\s+[a-zA-Z0-9._-]+)\s*[:=]\s*['\"]?([a-zA-Z0-9._~+/-]{8,})['\"]?",
                "category": "credentials",
            },
        ],
    },
    "geo": {
        "enabled": False,
        "allowed_countries": ["US", "CA", "GB", "DE"],
        "blocked_countries": [],
        "allowed_ip_ranges": [],
        "blocked_ip_ranges": [],
    },
    "time": {
        "enabled": False,
        "allowed_days": ["Mon", "Tue", "Wed", "Thu", "Fri"],
        "start_time": "08:00",
        "end_time": "18:00",
        "timezone": "UTC",
    },
}


def generate_totp(secret, timestamp=None, digits=6, step=30):
    """
    Generates an RFC 6238 TOTP token given a base32/string secret key.
    """
    if not secret:
        return ""
    if timestamp is None:
        timestamp = int(time.time())

    counter = int(timestamp) // step

    secret_clean = str(secret).upper().replace(" ", "").rstrip("=")
    missing_padding = len(secret_clean) % 8
    if missing_padding:
        secret_clean += "=" * (8 - missing_padding)

    try:
        key = base64.b32decode(secret_clean, casefold=True)
    except Exception:
        key = str(secret).encode("utf-8")

    msg = struct.pack(">Q", counter)
    h = hmac.new(key, msg, hashlib.sha1).digest()
    offset = h[-1] & 0x0F
    code = (struct.unpack(">I", h[offset : offset + 4])[0] & 0x7FFFFFFF) % (10**digits)
    return str(code).zfill(digits)


def verify_totp(token, secret, window=1, timestamp=None, digits=6, step=30):
    """
    Verifies a TOTP token against a secret key within a time window.
    """
    if token is None or not secret:
        return False
    token_str = str(token).strip()
    if not token_str:
        return False

    if timestamp is None:
        timestamp = int(time.time())

    for w in range(-window, window + 1):
        test_time = timestamp + (w * step)
        if generate_totp(secret, test_time, digits, step) == token_str:
            return True

    return False


def is_sensitive_command(cmd_obj_or_str, sensitive_keywords=None):
    """
    Determines if a command is sensitive based on flags, sensitivity level, or keywords.
    """
    if not cmd_obj_or_str:
        return False

    keywords = (
        sensitive_keywords
        if isinstance(sensitive_keywords, list)
        else DEFAULT_POLICY_CONFIG["mfa"]["sensitive_keywords"]
    )

    if isinstance(cmd_obj_or_str, dict):
        if (
            cmd_obj_or_str.get("sensitive") is True
            or cmd_obj_or_str.get("mfa_required") is True
        ):
            return True
        if cmd_obj_or_str.get("sensitivity_level") in ("high", "critical"):
            return True
        full_str = f"{cmd_obj_or_str.get('name', '')} {cmd_obj_or_str.get('command', '') or cmd_obj_or_str.get('template', '')}"
    else:
        full_str = str(cmd_obj_or_str)

    lower_str = full_str.lower()
    for kw in keywords:
        if kw and str(kw).strip():
            pattern = r"\b" + re.escape(str(kw).strip()) + r"\b"
            if re.search(pattern, lower_str):
                return True

    return False


def ip_in_range(ip_str, cidr_str):
    """
    Checks if an IP address is within a CIDR range or matches an exact IP.
    """
    if not ip_str or not cidr_str:
        return False
    try:
        ip = ipaddress.ip_address(str(ip_str).strip())
        net = ipaddress.ip_network(str(cidr_str).strip(), strict=False)
        return ip in net
    except Exception:
        return False


def check_geo(context, geo_policy=None):
    """
    Evaluates Geographic restrictions against context.
    Returns dict: {'allowed': bool, 'reason': str}
    """
    policy = geo_policy or DEFAULT_POLICY_CONFIG["geo"]
    if not policy or not policy.get("enabled", True):
        return {"allowed": True}

    ctx = context or {}
    ip = ctx.get("ip_address") or ctx.get("ip") or ""
    country = str(ctx.get("country_code") or ctx.get("country") or "").upper()

    allowed_countries = [str(c).upper() for c in policy.get("allowed_countries", [])]
    blocked_countries = [str(c).upper() for c in policy.get("blocked_countries", [])]
    allowed_ip_ranges = policy.get("allowed_ip_ranges", [])
    blocked_ip_ranges = policy.get("blocked_ip_ranges", [])

    if blocked_countries and country and country in blocked_countries:
        return {
            "allowed": False,
            "reason": f"Country '{country}' is explicitly blocked by geographic policy.",
        }

    if allowed_countries and country and country not in allowed_countries:
        return {
            "allowed": False,
            "reason": f"Country '{country}' is not in allowed geographic regions list.",
        }

    if ip and blocked_ip_ranges:
        for r in blocked_ip_ranges:
            if ip_in_range(ip, r):
                return {
                    "allowed": False,
                    "reason": f"IP address '{ip}' is blocked by IP range policy ({r}).",
                }

    if ip and allowed_ip_ranges:
        in_allowed = any(ip_in_range(ip, r) for r in allowed_ip_ranges)
        if not in_allowed:
            return {
                "allowed": False,
                "reason": f"IP address '{ip}' is not in allowed IP ranges list.",
            }

    return {"allowed": True}


def check_time(context, time_policy=None):
    """
    Evaluates Time-based access controls against context.
    Returns dict: {'allowed': bool, 'reason': str}
    """
    policy = time_policy or DEFAULT_POLICY_CONFIG["time"]
    if not policy or not policy.get("enabled", True):
        return {"allowed": True}

    ctx = context or {}
    ts = ctx.get("timestamp")
    if isinstance(ts, (int, float)):
        now_dt = datetime.datetime.fromtimestamp(ts, tz=datetime.timezone.utc)
    elif isinstance(ts, datetime.datetime):
        now_dt = ts
    else:
        now_dt = datetime.datetime.now(datetime.timezone.utc)

    days_map = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
    current_day_name = days_map[now_dt.weekday()]
    current_day_idx = now_dt.weekday()  # Mon=0, Sun=6

    allowed_days = policy.get("allowed_days", [])
    if allowed_days:
        is_allowed = False
        for d in allowed_days:
            if isinstance(d, int) and d == current_day_idx:
                is_allowed = True
                break
            if isinstance(d, str) and d.lower()[:3] == current_day_name.lower():
                is_allowed = True
                break
        if not is_allowed:
            return {
                "allowed": False,
                "reason": f"Access denied on {current_day_name} (allowed days: {allowed_days}).",
            }

    start_time_str = policy.get("start_time")
    end_time_str = policy.get("end_time")

    if start_time_str and end_time_str:

        def parse_minutes(t_str):
            parts = [int(p) for p in str(t_str).split(":")]
            return parts[0] * 60 + parts[1]

        start_min = parse_minutes(start_time_str)
        end_min = parse_minutes(end_time_str)
        current_min = now_dt.hour * 60 + now_dt.minute

        if start_min <= end_min:
            if current_min < start_min or current_min > end_min:
                return {
                    "allowed": False,
                    "reason": f"Access denied at current time ({now_dt.strftime('%H:%M')} UTC). Allowed window: {start_time_str} - {end_time_str}.",
                }
        else:
            # Overnight window
            if current_min < start_min and current_min > end_min:
                return {
                    "allowed": False,
                    "reason": f"Access denied at current time ({now_dt.strftime('%H:%M')} UTC). Allowed overnight window: {start_time_str} - {end_time_str}.",
                }

    return {"allowed": True}


def check_dlp(text, dlp_policy=None):
    """
    Scans text for DLP pattern violations.
    Returns list of dicts: [{'name': str, 'category': str, 'match': str}]
    """
    if not text or not isinstance(text, str):
        return []

    policy = dlp_policy or DEFAULT_POLICY_CONFIG["dlp"]
    if not policy or not policy.get("enabled", True):
        return []

    patterns = policy.get("patterns") or DEFAULT_POLICY_CONFIG["dlp"]["patterns"]
    violations = []

    for pat in patterns:
        if not pat or not pat.get("regex"):
            continue
        try:
            raw_regex = pat["regex"]
            flags = 0
            if raw_regex.startswith("(?i)"):
                flags = re.IGNORECASE
                raw_regex = raw_regex[4:]

            matches = re.finditer(raw_regex, text, flags=flags)
            for m in matches:
                violations.append(
                    {
                        "name": pat.get("name", "Sensitive Data"),
                        "category": pat.get("category", "dlp"),
                        "match": m.group(0),
                    }
                )
        except Exception:
            pass

    return violations


def redact_dlp(text, dlp_policy=None):
    """
    Redacts sensitive DLP pattern matches in text.
    """
    if not text or not isinstance(text, str):
        return ""

    policy = dlp_policy or DEFAULT_POLICY_CONFIG["dlp"]
    if not policy or not policy.get("enabled", True):
        return text

    patterns = policy.get("patterns") or DEFAULT_POLICY_CONFIG["dlp"]["patterns"]
    redacted = text

    for pat in patterns:
        if not pat or not pat.get("regex"):
            continue
        try:
            raw_regex = pat["regex"]
            flags = 0
            if raw_regex.startswith("(?i)"):
                flags = re.IGNORECASE
                raw_regex = raw_regex[4:]

            replacement = f"[REDACTED_{pat.get('category', 'DATA').upper()}]"
            redacted = re.sub(raw_regex, replacement, redacted, flags=flags)
        except Exception:
            pass

    return redacted


class PolicyEngine:
    """
    Stateful and configurable Policy Enforcement Engine class.
    """

    def __init__(self, policy_config=None):
        self.policy = policy_config or DEFAULT_POLICY_CONFIG

    def update_policy(self, new_policy):
        self.policy = new_policy

    def evaluate_policy(self, command_obj_or_str, params=None, context=None):
        return evaluate_policy(command_obj_or_str, params, context, self.policy)


def evaluate_policy(command_obj_or_str, params=None, context=None, policy_config=None):
    """
    Evaluates Security Policy for a command execution.
    Evaluation order: Time -> Geo -> DLP -> MFA.
    """
    policy = policy_config or DEFAULT_POLICY_CONFIG
    ctx = context or {}
    params_dict = dict(params) if isinstance(params, dict) else {}

    if isinstance(command_obj_or_str, dict):
        cmd_obj = command_obj_or_str
    else:
        cmd_obj = {
            "command": str(command_obj_or_str or ""),
            "name": str(command_obj_or_str or ""),
        }

    raw_cmd_str = cmd_obj.get("command") or cmd_obj.get("template") or ""
    param_str = " ".join(str(v) for v in params_dict.values())
    full_text_to_scan = f"{cmd_obj.get('name', '')} {raw_cmd_str} {param_str}".strip()

    reasons = []
    dlp_violations = []

    if not policy.get("enabled", True):
        return {
            "allowed": True,
            "action": "allow",
            "reasons": ["Policy enforcement is globally disabled."],
            "mfa_required": False,
            "mfa_verified": False,
            "dlp_violations": [],
            "sanitized_command": raw_cmd_str,
            "sanitized_params": params_dict,
            "evaluated_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        }

    # 1. Time Policy Check
    time_res = check_time(ctx, policy.get("time"))
    if not time_res["allowed"]:
        return {
            "allowed": False,
            "action": "block",
            "reasons": [
                time_res.get("reason", "Time-based policy restriction violation.")
            ],
            "mfa_required": False,
            "mfa_verified": False,
            "dlp_violations": [],
            "sanitized_command": raw_cmd_str,
            "sanitized_params": params_dict,
            "evaluated_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        }

    # 2. Geo Policy Check
    geo_res = check_geo(ctx, policy.get("geo"))
    if not geo_res["allowed"]:
        return {
            "allowed": False,
            "action": "block",
            "reasons": [
                geo_res.get("reason", "Geographic policy restriction violation.")
            ],
            "mfa_required": False,
            "mfa_verified": False,
            "dlp_violations": [],
            "sanitized_command": raw_cmd_str,
            "sanitized_params": params_dict,
            "evaluated_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        }

    # 3. DLP Policy Check
    sanitized_cmd = raw_cmd_str
    sanitized_params = dict(params_dict)

    dlp_policy = policy.get("dlp", {})
    if dlp_policy.get("enabled", True):
        found_violations = check_dlp(full_text_to_scan, dlp_policy)
        if found_violations:
            dlp_violations.extend(found_violations)
            action = dlp_policy.get("action", "block")

            if action == "block":
                v_names = ", ".join(set(v["name"] for v in found_violations))
                return {
                    "allowed": False,
                    "action": "block",
                    "reasons": [
                        f"Data Loss Prevention (DLP) violation: {v_names} detected."
                    ],
                    "mfa_required": False,
                    "mfa_verified": False,
                    "dlp_violations": found_violations,
                    "sanitized_command": raw_cmd_str,
                    "sanitized_params": params_dict,
                    "evaluated_at": datetime.datetime.now(
                        datetime.timezone.utc
                    ).isoformat(),
                }
            elif action == "redact":
                sanitized_cmd = redact_dlp(raw_cmd_str, dlp_policy)
                for k, v in sanitized_params.items():
                    sanitized_params[k] = redact_dlp(str(v), dlp_policy)
                reasons.append(
                    f"DLP policy automatically redacted sensitive content ({len(found_violations)} items)."
                )
            elif action == "warn":
                v_names = ", ".join(set(v["name"] for v in found_violations))
                reasons.append(f"DLP warning: sensitive content detected ({v_names}).")

    # 4. MFA Policy Check
    mfa_required = False
    mfa_verified = False

    mfa_policy = policy.get("mfa", {})
    if mfa_policy.get("enabled", True):
        mfa_required = is_sensitive_command(
            cmd_obj, mfa_policy.get("sensitive_keywords")
        )

        if mfa_required:
            token = ctx.get("mfa_token") or params_dict.get("mfa_token")
            session_token = ctx.get("session_token")
            secret = mfa_policy.get("secret") or DEFAULT_POLICY_CONFIG["mfa"]["secret"]

            if session_token and ctx.get("session_valid") is True:
                mfa_verified = True
                reasons.append("MFA verified via active session.")
            elif token and verify_totp(token, secret):
                mfa_verified = True
                reasons.append("MFA verified via TOTP code.")
            else:
                return {
                    "allowed": False,
                    "action": "mfa_required",
                    "reasons": [
                        "Operation is classified as sensitive and requires Multi-Factor Authentication (MFA)."
                    ],
                    "mfa_required": True,
                    "mfa_verified": False,
                    "dlp_violations": dlp_violations,
                    "sanitized_command": sanitized_cmd,
                    "sanitized_params": sanitized_params,
                    "evaluated_at": datetime.datetime.now(
                        datetime.timezone.utc
                    ).isoformat(),
                }

    if not reasons:
        reasons.append("Policy evaluation passed successfully.")

    return {
        "allowed": True,
        "action": (
            "redact"
            if (dlp_violations and dlp_policy.get("action") == "redact")
            else "allow"
        ),
        "reasons": reasons,
        "mfa_required": mfa_required,
        "mfa_verified": mfa_verified,
        "dlp_violations": dlp_violations,
        "sanitized_command": sanitized_cmd,
        "sanitized_params": sanitized_params,
        "evaluated_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
    }

File: extension/policyEngine.js
/**
 * Policy Enforcement Engine for CmdBar extension.
 * Supports:
 * - Multi-Factor Authentication (MFA) for sensitive operations
 * - Data Loss Prevention (DLP) for inputs and outputs (Credit Cards, SSN, API keys, Private keys)
 * - Geographic restrictions (Country code allow/deny lists, IP CIDR matching)
 * - Time-based access control (Allowed days, allowed hour ranges, timezone support)
 */

export const DEFAULT_DLP_PATTERNS = {
  credit_card: {
    name: "Credit Card Number",
    regex: /\b(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13}|6(?:011|5[0-9]{2})[0-9]{12})\b/g,
    redactLabel: "[REDACTED:CREDIT_CARD]",
    luhnCheck: true,
  },
  ssn: {
    name: "Social Security Number",
    regex: /\b\d{3}-\d{2}-\d{4}\b/g,
    redactLabel: "[REDACTED:SSN]",
  },
  aws_key: {
    name: "AWS Access Key",
    regex: /\b(AKIA|ASIA)[0-9A-Z]{16}\b/g,
    redactLabel: "[REDACTED:AWS_KEY]",
  },
  private_key: {
    name: "Private Key Header",
    regex: /-----BEGIN (?:RSA|OPENSSH|EC|DSA|PGP)? PRIVATE KEY-----[\s\S]*?-----END (?:RSA|OPENSSH|EC|DSA|PGP)? PRIVATE KEY-----/g,
    redactLabel: "[REDACTED:PRIVATE_KEY]",
  },
  api_token: {
    name: "API Token / Password",
    regex: /\b(?:bearer|api[-_]?key|secret|password|passwd|auth[-_]?token)\b\s*[:=]\s*["']?([A-Za-z0-9_\-\.\~]{12,})["']?/gi,
    redactLabel: "[REDACTED:API_TOKEN]",
  },
};

export const DEFAULT_SENSITIVE_COMMAND_PATTERNS = [
  "*deploy*",
  "*rm -rf*",
  "*sudo*",
  "*aws ecs*",
  "*db:drop*",
  "*db:migrate*",
  "*pkill*",
];

/**
 * Validates a number string using the Luhn Algorithm.
 * @param {string} numStr
 * @returns {boolean}
 */
export function validateLuhn(numStr) {
  const clean = numStr.replace(/\D/g, "");
  if (clean.length < 11 || clean.length > 19) return false;

  let sum = 0;
  let shouldDouble = false;
  for (let i = clean.length - 1; i >= 0; i--) {
    let digit = parseInt(clean.charAt(i), 10);
    if (shouldDouble) {
      digit *= 2;
      if (digit > 9) digit -= 9;
    }
    sum += digit;
    shouldDouble = !shouldDouble;
  }
  return sum % 10 === 0;
}

/**
 * Simple IPv4 in CIDR range checker.
 * @param {string} ip
 * @param {string} cidr
 * @returns {boolean}
 */
export function isIpInCidr(ip, cidr) {
  if (!ip || !cidr) return false;
  if (ip === cidr) return true;

  if (!cidr.includes("/")) {
    return ip === cidr;
  }

  const [rangeIp, prefixStr] = cidr.split("/");
  const prefix = parseInt(prefixStr, 10);
  if (isNaN(prefix) || prefix < 0 || prefix > 32) return false;

  const ipNum = ipToLong(ip);
  const rangeNum = ipToLong(rangeIp);

  if (ipNum === null || rangeNum === null) return false;

  const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
  return (ipNum & mask) === (rangeNum & mask);
}

function ipToLong(ip) {
  const parts = ip.trim().split(".");
  if (parts.length !== 4) return null;
  let num = 0;
  for (let i = 0; i < 4; i++) {
    const part = parseInt(parts[i], 10);
    if (isNaN(part) || part < 0 || part > 255) return null;
    num = (num << 8) + part;
  }
  return num >>> 0;
}

export class PolicyEngine {
  /**
   * @param {object} [policyConfig]
   */
  constructor(policyConfig = {}) {
    this.mfaSession = {
      verified: false,
      timestamp: 0,
      token: null,
    };

    this.configure(policyConfig);
  }

  /**
   * Updates or replaces the policy engine configuration.
   * @param {object} policyConfig
   */
  configure(policyConfig = {}) {
    this.mfaConfig = {
      enabled: policyConfig.mfa?.enabled ?? false,
      sensitiveCommands: policyConfig.mfa?.sensitiveCommands || [
        ...DEFAULT_SENSITIVE_COMMAND_PATTERNS,
      ],
      mfaSecret: policyConfig.mfa?.mfaSecret || "123456",
      sessionDurationMs: policyConfig.mfa?.sessionDurationMs || 300000, // 5 minutes default
    };

    this.dlpConfig = {
      enabled: policyConfig.dlp?.enabled ?? true,
      action: policyConfig.dlp?.action || "redact", // "redact", "block", or "warn"
      customPatterns: policyConfig.dlp?.customPatterns || {},
    };

    this.geoConfig = {
      enabled: policyConfig.geo?.enabled ?? false,
      allowedCountries: policyConfig.geo?.allowedCountries || [],
      blockedCountries: policyConfig.geo?.blockedCountries || [],
      allowedIpRanges: policyConfig.geo?.allowedIpRanges || [],
      blockedIpRanges: policyConfig.geo?.blockedIpRanges || [],
    };

    this.timeConfig = {
      enabled: policyConfig.time?.enabled ?? false,
      allowedDays: policyConfig.time?.allowedDays || [
        "Mon",
        "Tue",
        "Wed",
        "Thu",
        "Fri",
      ],
      allowedHours: policyConfig.time?.allowedHours || {
        start: "00:00",
        end: "23:59",
      },
      timezone: policyConfig.time?.timezone || "UTC",
    };
  }

  /**
   * Checks if an MFA session is currently active and valid.
   * @param {number} [nowTimestamp]
   * @returns {boolean}
   */
  isMfaSessionValid(nowTimestamp = Date.now()) {
    if (!this.mfaSession.verified) return false;
    const elapsed = nowTimestamp - this.mfaSession.timestamp;
    return elapsed >= 0 && elapsed <= this.mfaConfig.sessionDurationMs;
  }

  /**
   * Verifies an MFA token against secret or validator.
   * @param {string} token
   * @param {number} [nowTimestamp]
   * @returns {boolean}
   */
  verifyMfaToken(token, nowTimestamp = Date.now()) {
    if (!token) return false;
    const cleanToken = String(token).trim();
    if (cleanToken === String(this.mfaConfig.mfaSecret).trim()) {
      this.mfaSession = {
        verified: true,
        timestamp: nowTimestamp,
        token: cleanToken,
      };
      return true;
    }
    return false;
  }

  /**
   * Clears active MFA session.
   */
  clearMfaSession() {
    this.mfaSession = {
      verified: false,
      timestamp: 0,
      token: null,
    };
  }

  /**
   * Determines if a command or operation is considered sensitive and requires MFA.
   * @param {object|string} commandObj
   * @returns {boolean}
   */
  isSensitiveOperation(commandObj) {
    if (!commandObj) return false;

    let cmdStr = "";
    let isExplicitSensitive = false;

    if (typeof commandObj === "string") {
      cmdStr = commandObj;
    } else if (typeof commandObj === "object") {
      cmdStr = commandObj.command || commandObj.name || "";
      if (commandObj.sensitive || commandObj.require_mfa) {
        isExplicitSensitive = true;
      }
    }

    if (isExplicitSensitive) return true;

    const lowerCmd = cmdStr.toLowerCase().trim();
    for (const pattern of this.mfaConfig.sensitiveCommands) {
      const p = pattern.toLowerCase();
      if (p.startsWith("*") && p.endsWith("*")) {
        const sub = p.slice(1, -1);
        if (lowerCmd.includes(sub)) return true;
      } else if (p.startsWith("*")) {
        const sub = p.slice(1);
        if (lowerCmd.endsWith(sub)) return true;
      } else if (p.endsWith("*")) {
        const sub = p.slice(0, -1);
        if (lowerCmd.startsWith(sub)) return true;
      } else if (lowerCmd === p) {
        return true;
      }
    }

    return false;
  }

  /**
   * Scans text for sensitive data loss patterns (DLP).
   * @param {string} text
   * @returns {{ matches: Array<{ rule: string, value: string }>, redactedText: string }}
   */
  scanDlp(text) {
    if (!text || typeof text !== "string") {
      return { matches: [], redactedText: text || "" };
    }

    const matches = [];
    let redactedText = text;

    const allPatterns = {
      ...DEFAULT_DLP_PATTERNS,
      ...this.dlpConfig.customPatterns,
    };

    for (const [ruleKey, rule] of Object.entries(allPatterns)) {
      const regex = new RegExp(rule.regex.source, rule.regex.flags || "g");
      let match;
      while ((match = regex.exec(text)) !== null) {
        const val = match[0];
        if (rule.luhnCheck) {
          if (!validateLuhn(val)) continue;
        }

        matches.push({
          rule: ruleKey,
          name: rule.name || ruleKey,
          value: val,
        });

        redactedText = redactedText.replace(
          val,
          rule.redactLabel || "[DLP REDACTED]",
        );
      }
    }

    return { matches, redactedText };
  }

  /**
   * Evaluates geographic restrictions based on user location context.
   * @param {object} geoContext
   * @returns {{ allowed: boolean, reason: string|null }}
   */
  evaluateGeo(geoContext = {}) {
    if (!this.geoConfig.enabled) {
      return { allowed: true, reason: null };
    }

    const country = (geoContext.countryCode || geoContext.country || "").toUpperCase().trim();
    const ip = (geoContext.ip || "").trim();

    // 1. Country code check
    if (country) {
      if (
        this.geoConfig.blockedCountries.length > 0 &&
        this.geoConfig.blockedCountries.map((c) => c.toUpperCase()).includes(country)
      ) {
        return {
          allowed: false,
          reason: `Geographic Restriction: Country '${country}' is blocked.`,
        };
      }

      if (
        this.geoConfig.allowedCountries.length > 0 &&
        !this.geoConfig.allowedCountries.map((c) => c.toUpperCase()).includes(country)
      ) {
        return {
          allowed: false,
          reason: `Geographic Restriction: Country '${country}' is not in allowed list.`,
        };
      }
    }

    // 2. IP Range check
    if (ip) {
      if (this.geoConfig.blockedIpRanges.length > 0) {
        for (const cidr of this.geoConfig.blockedIpRanges) {
          if (isIpInCidr(ip, cidr)) {
            return {
              allowed: false,
              reason: `Geographic Restriction: IP address '${ip}' matches blocked range '${cidr}'.`,
            };
          }
        }
      }

      if (this.geoConfig.allowedIpRanges.length > 0) {
        let matched = false;
        for (const cidr of this.geoConfig.allowedIpRanges) {
          if (isIpInCidr(ip, cidr)) {
            matched = true;
            break;
          }
        }
        if (!matched) {
          return {
            allowed: false,
            reason: `Geographic Restriction: IP address '${ip}' is not in allowed IP ranges.`,
          };
        }
      }
    }

    return { allowed: true, reason: null };
  }

  /**
   * Evaluates time-based access control against execution timestamp.
   * @param {Date|number|string} [timeContext]
   * @returns {{ allowed: boolean, reason: string|null }}
   */
  evaluateTime(timeContext) {
    if (!this.timeConfig.enabled) {
      return { allowed: true, reason: null };
    }

    const dateObj = timeContext ? new Date(timeContext) : new Date();
    if (isNaN(dateObj.getTime())) {
      return { allowed: true, reason: null };
    }

    const dayNames = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
    const currentDay = dayNames[dateObj.getUTCDay()];

    if (
      this.timeConfig.allowedDays &&
      this.timeConfig.allowedDays.length > 0
    ) {
      const allowedDaysUpper = this.timeConfig.allowedDays.map((d) =>
        String(d).toUpperCase().slice(0, 3),
      );
      if (!allowedDaysUpper.includes(currentDay.toUpperCase())) {
        return {
          allowed: false,
          reason: `Time Restriction: Access not allowed on ${currentDay}. Allowed days: ${this.timeConfig.allowedDays.join(
            ", ",
          )}.`,
        };
      }
    }

    if (this.timeConfig.allowedHours) {
      const startStr = this.timeConfig.allowedHours.start || "00:00";
      const endStr = this.timeConfig.allowedHours.end || "23:59";

      const [startH, startM] = startStr.split(":").map((v) => parseInt(v, 10) || 0);
      const [endH, endM] = endStr.split(":").map((v) => parseInt(v, 10) || 0);

      const currentMinutes = dateObj.getUTCHours() * 60 + dateObj.getUTCMinutes();
      const startMinutes = startH * 60 + startM;
      const endMinutes = endH * 60 + endM;

      if (currentMinutes < startMinutes || currentMinutes > endMinutes) {
        const timeStr = `${String(dateObj.getUTCHours()).padStart(2, "0")}:${String(
          dateObj.getUTCMinutes(),
        ).padStart(2, "0")}`;
        return {
          allowed: false,
          reason: `Time Restriction: Access not allowed at ${timeStr} UTC. Allowed hours: ${startStr} - ${endStr}.`,
        };
      }
    }

    return { allowed: true, reason: null };
  }

  /**
   * Main evaluation entry point for command execution contexts.
   * @param {object} evalContext
   * @param {object|string} evalContext.command
   * @param {object} [evalContext.parameters]
   * @param {string} [evalContext.mfaToken]
   * @param {object} [evalContext.geo]
   * @param {Date|number|string} [evalContext.timestamp]
   * @returns {object} Evaluation results
   */
  evaluate(evalContext = {}) {
    const reasons = [];
    let allowed = true;

    const now = evalContext.timestamp
      ? new Date(evalContext.timestamp).getTime()
      : Date.now();

    // 1. Geographic Check
    const geoRes = this.evaluateGeo(evalContext.geo || {});
    if (!geoRes.allowed) {
      allowed = false;
      reasons.push(geoRes.reason);
    }

    // 2. Time Check
    const timeRes = this.evaluateTime(evalContext.timestamp);
    if (!timeRes.allowed) {
      allowed = false;
      reasons.push(timeRes.reason);
    }

    // 3. MFA Check for Sensitive Operations
    const isSensitive = this.isSensitiveOperation(evalContext.command);
    let mfaRequired = false;
    let mfaVerified = false;

    if (this.mfaConfig.enabled && isSensitive) {
      mfaRequired = true;
      if (evalContext.mfaToken) {
        this.verifyMfaToken(evalContext.mfaToken, now);
      }
      mfaVerified = this.isMfaSessionValid(now);

      if (!mfaVerified) {
        allowed = false;
        reasons.push("MFA Authentication Required: Sensitive operation requires verified MFA token.");
      }
    }

    // 4. DLP Scan on Inputs
    let redactedCommand =
      typeof evalContext.command === "string"
        ? evalContext.command
        : evalContext.command?.command || "";
    let redactedParameters = { ...(evalContext.parameters || {}) };
    let dlpMatches = [];

    if (this.dlpConfig.enabled) {
      const cmdScan = this.scanDlp(redactedCommand);
      dlpMatches.push(...cmdScan.matches);
      redactedCommand = cmdScan.redactedText;

      for (const [key, val] of Object.entries(redactedParameters)) {
        if (typeof val === "string") {
          const paramScan = this.scanDlp(val);
          dlpMatches.push(...paramScan.matches);
          redactedParameters[key] = paramScan.redactedText;
        }
      }

      if (dlpMatches.length > 0) {
        if (this.dlpConfig.action === "block") {
          allowed = false;
          reasons.push(`DLP Violation: Blocked due to sensitive data detection (${dlpMatches.map((m) => m.name).join(", ")}).`);
        } else if (this.dlpConfig.action === "warn") {
          reasons.push(`DLP Warning: Sensitive data detected (${dlpMatches.map((m) => m.name).join(", ")}).`);
        }
      }
    }

    return {
      allowed,
      mfaRequired,
      mfaVerified,
      reasons,
      dlpMatches,
      redactedCommand,
      redactedParameters,
      redactOutput: (outputStr) => {
        if (!this.dlpConfig.enabled || !outputStr) return outputStr;
        const scan = this.scanDlp(outputStr);
        return scan.redactedText;
      },
    };
  }
}

export function createPolicyEngine(config = {}) {
  return new PolicyEngine(config);
}
